README.md
KEGS is an Apple IIgs emulator for Mac, Linux, and Windows.

You must provide a ROM file (supports just about any known format, ROM01 or ROM03), and bootable image.  The directions provide details on what you need.

Press F4 while it is running to change options.

KEGS is likely the fastest IIgs emulator, hitting over 1GHz on Mac M1 systems.  It has many features, such as the ability to mount disk images directly from .zip files.

Binaries and other files are at:  https://kegs.sourceforge.net/


doc/CHANGES.txt

Changes in KEGS v1.33 since v1.32 (12/10/23)
- Add command line argument support to set any disk (-s5d1=Data.po
	or -s7d12 bigdisk.hdv) or any knob that is listed in config.kegs.
- Add -cfg path_and_name_of_config_kegs file to use a particular config.kegs
	file.
- Fix Code Red when leaving the configuration screen (F4) using the last
	menu item "Exit Config".
- Allow pasting of control-characters such as Ctrl-D and Ctrl-H.


Changes in KEGS v1.32 since v1.31 (11/22/23)
- Fix (dloc,x) in emulation mode to wrap as described at
	https://github.com/gilyon/snes-tests/tree/main/cputest
- Improve virtual hard drive in slot 7 to use a small driver at $C700 which
	uses WDM $C7,$00 to call back to KEGS for handling commands.
- Improve SCC8530 emulation to add support for RTxC as clock for higher speeds.
- Fix a GS/OS visual anomaly with the mouse cursor sometimes disappearing
	when moving upwards, due to Scanline interrupts not being taken
	properly.
- Support qkumba's code to run from $C050 by having get_remaing_opcodes()
	track time properly, and to have reads to $C050-$C057 call float_bus()
	before doing the softswitch action.


Changes in KEGS v1.31 since v1.30 (11/04/23)
- Fix Windows failure where KEGS would quit on startup if config.kegs
	contained a new ROM path.
- Fix a Code Red halt running the Printer57.6 driver where KEGS thought it
	might need to generate a baud rate event every .5 cycles.
- Fix disk image selection screen bug where s7d10-s7d12 could wrap and make
	it hard to leave the screen.
- Add a Slinky RAM card in slot 4 (with no firmware), works even with
	Mockingboard.
- Fix scanline interrupts which were happening too early starting with
	version 1.24.
- Another false read bug was causing 16-bit RMW cycles to read the next
	address (which is incorrect).


Changes in KEGS v1.30 since v1.29 (09/23/23)
- Proper emulation of the $C080-$C08F language card soft switches.
- Improved INTC8ROM emulation, so a2audit passes (with Apple //e ROMs)
- Fix SCC RR2B register emulation for Colin Leroy-Mira's telnet.system.
- Fix SCC remote IP mode to restart the connection if the remote side ended
	it due to idleness, to better handle connecting to printers.
- RAMRD/RAMWRT/ST80COL/etc. apply to bank $E0 as well as to bank $00.  This
	was not emulated properly before.  This fixes the AppleLink Terminal
	application.


Changes in KEGS v1.29 since v1.28 (09/05/23)
- Improved disk arm emulation for 5.25" disks
- Enable use of a real serial port on Linux, and improve real serial port
	emulation on Mac.
- Add serial "outgoing IP" to allow slot 1 serial to be sent directly to
	a real printer (often port 9100).
- When mounting an image from a .zip file, you can press Cmd-A and all
	subsequent images will also be mounted in consecutive drives.
	This is useful for the new wita2gs_0_70.zip.
- Allow "unlocking" locked images from .zip files, to allow code to write to
	them (but it's all in memory, so all changes are lost when KEGS
	exits).
- Fix DiskCopy4.2 image detection, to handle images of sizes other than 800KB.


Changes in KEGS v1.28 since v1.27 (06/21/23)
- Reduce status lines under the window from 7 lines of text to 4.
- Fix Windows10 crash when KEGS was minimized (Windows set the window size to
	0, which was unexpected, and led to a divide-by-0).
- Enable live window resizing on Windows64.  Speed up the video scaling for
	X11 and Windows.
- Allow the ZipGS speed, which was fixed at 8MHz previously), be set to
	8MHz, 16MHz, 32MHz, 64MHz, or 128MHz.
- Fix a false read bug which broke SCC emulation.  LDA $BFFD,X where X=$3D
	was "false" reading $C039, not $BF39 as it should, leading to SCC
	state being incorrect.


Changes in KEGS v1.27 since v1.26 (06/13/23)
- Ignore WDM 0xfc,0xfd,0xff to avoid HOST.FST causing Code Red.
- Fix $C019 reading to match Deater's Midline demo
- Add false reads for RMW instruction, (dloc),y, abs,x, and abs,y modes.
	Only 8-bit false reads are done currently (this only affect 16-bit
	RMW, where I know of nothing using the false reads).
- Fix reported bug where long paths in the file selection screens didn't
	truncate the files so the endings could always be seen.
- Allow 140K .SDK images to be selected in the image file selection screen.
- Fix $C030 speaker toggle emulation to eliminate a 60-Hertz buzz caused by
	the code being organized around 60Hz screen refresh and not counting
	on the last toggle correctly in each video frame in all cases.  The
	$C030 speaker output ramps down to 0 after about 60msec to avoid the
	annoying "click" 4 seconds later when KEGS pauses sound output.


Changes in KEGS v1.26 since v1.25 (05/22/23)
- Fix Win64 Dynapro issues (O_BINARY, setvbuf was causing a crash).
- Fix a KEGS bug in the Bank $E0 memory map which could corrupt data in the
	Apple IIgs memory from $E0/6000 - $E0/A000 introduced in KEGS 1.20.


Changes in KEGS v1.25 since v1.24 (05/21/23)
- Actual Win32 support.  1.24 was only Win64.


Changes in KEGS v1.24 since v1.23 (05/17/23)
- Win64 support.  kegswin.exe now part of the standard release.  The Windows
	port is still beta quality.
- Try to fix jerky video to make KEGS seem smoother.


Changes in KEGS v1.23 since v1.22 (05/05/23)
- Change the way KEGS tracks time from a double to a unsigned long long,
	which enables higher speeds.
- Support video mode changes in the middle of lines.


Changes in KEGS v1.22 since v1.21 (04/27/23)
- Remove debugging printfs from iwm.c.
- Fix the way video updates are done to fix Dagen Brock's HDGR demo (which
	switches pages to double the GR vertical resolution).


Changes in KEGS v1.21 since v1.20 (04/15/23)
- Fix Antoine's reported issue where KEGS would hang after ejecting 3.5" disks.
	Caused by a debug statement accidentally left in for the release.
- Fix Stephan's reported issue with keys repeating forever with French
	keyboards.


Changes in KEGS v1.20 since v1.19 (03/31/23)
- Ctrl-F9 is now Copy.  The text screen is copied to your host system

src/clock.c
const char rcsid_clock_c[] = "@(#)$KmKId: clock.c,v 1.38 2023-05-19 13:52:30+00 kentd Exp $";

/************************************************************************/
/*			KEGS: Apple //gs Emulator			*/
/*			Copyright 2002-2022 by Kent Dickey		*/
/*									*/
/*	This code is covered by the GNU GPL v3				*/
/*	See the file COPYING.txt or https://www.gnu.org/licenses/	*/
/*	This program is provided with no warranty			*/
/*									*/
/*	The KEGS web page is kegs.sourceforge.net			*/
/*	You may contact the author at: kadickey@alumni.princeton.edu	*/
/************************************************************************/

#include "defc.h"
#include <time.h>
#ifdef _WIN32
# include <windows.h>
# include <mmsystem.h>
#else
# include <sys/time.h>
#endif

extern int Verbose;
extern int g_vbl_count;
extern int g_rom_version;
extern int g_config_kegs_update_needed;

#define CLK_IDLE		1
#define CLK_TIME		2
#define CLK_INTERNAL		3
#define CLK_BRAM1		4
#define CLK_BRAM2		5

int	g_clk_mode = CLK_IDLE;
int	g_clk_read = 0;
int	g_clk_reg1 = 0;

extern int g_c033_data;
extern int g_c034_val;

byte	g_bram[2][256];
byte	*g_bram_ptr = &(g_bram[0][0]);

word32 g_clk_cur_time = 0xa0000000;
int	g_clk_next_vbl_update = 0;

double
get_dtime()
{
#ifndef _WIN32
	struct timeval tp1;
	double	dsec;
	double	dusec;
#endif
	double	dtime;

	/* Routine used to return actual system time as a double */
	/* No routine cares about the absolute value, only deltas--maybe */
	/*  take advantage of that in future to increase usec accuracy */

#ifdef _WIN32
	dtime = timeGetTime() / 1000.0;
#else

# ifdef SOLARIS
	gettimeofday(&tp1, (void *)0);
# else
	gettimeofday(&tp1, (struct timezone *)0);
# endif

	dsec = (double)tp1.tv_sec;
	dusec = (double)tp1.tv_usec;

	dtime = dsec + (dusec / (1000.0 * 1000.0));
#endif

	return dtime;
}

int
micro_sleep(double dtime)
{
#ifndef _WIN32
	struct timeval Timer;
	int	ret;
#endif

	if(dtime <= 0.0) {
		return 0;
	}
	if(dtime >= 1.0) {
		halt_printf("micro_sleep called with %f!!\n", dtime);
		return -1;
	}

#if 0
	printf("usleep: %f\n", dtime);
#endif

#ifdef _WIN32
	Sleep((word32)(dtime * 1000));
#else
	Timer.tv_sec = 0;
	Timer.tv_usec = (dtime * 1000000.0);
	if( (ret = select(0, 0, 0, 0, &Timer)) < 0) {
		fprintf(stderr, "micro_sleep (select) ret: %d, errno: %d\n",
			ret, errno);
		return -1;
	}
#endif
	return 0;
}

void
clk_bram_zero()
{
	int	i, j;

	/* zero out all bram */
	for(i = 0; i < 2; i++) {
		for(j = 0; j < 256; j++) {
			g_bram[i][j] = 0;
		}
	}
	g_bram_ptr = &(g_bram[0][0]);
}

void
clk_bram_set(int bram_num, int offset, int val)
{
	if((bram_num < 0) || (bram_num > 2)) {
		printf("bram_num %d out of range\n", bram_num);
		return;
	}
	if((offset < 0) || (offset > 0x100)) {
		printf("bram offset %05x out of range\n", offset);
		return;
	}
	g_bram[bram_num][offset] = val;
}

void
clk_setup_bram_version()
{
	if(g_rom_version < 3) {
		g_bram_ptr = (&g_bram[0][0]);	// ROM 01
	} else {
		g_bram_ptr = (&g_bram[1][0]);	// ROM 03
	}
}

void
clk_write_bram(FILE *fconf)
{
	int	i, j, k;

	for(i = 0; i < 2; i++) {
		fprintf(fconf, "\n");
		for(j = 0; j < 256; j += 16) {
			fprintf(fconf, "bram%d[%02x] =", 2*i + 1, j);
			for(k = 0; k < 16; k++) {
				fprintf(fconf, " %02x", g_bram[i][j+k]);
			}
			fprintf(fconf, "\n");
		}
	}
}

void
update_cur_time()
{
	struct tm *tm_ptr;
	time_t	cur_time, secs, secs2;

	cur_time = time(0);

	/* Figure out the timezone (effectively) by diffing two times. */
	/* this is probably not right for a few hours around daylight savings*/
	/*  time transition */
	secs2 = mktime(gmtime(&cur_time));
	tm_ptr = localtime(&cur_time);
	secs = mktime(tm_ptr);

	secs2 = secs2 - secs;		// this is the timezone offset
#ifdef MAC
	/* Mac OS X's mktime function modifies the tm_ptr passed in for */
	/*  the CDT timezone and breaks this algorithm.  So on a Mac, we */
	/*  will use the tm_ptr->gmtoff member to correct the time */
	secs = secs + tm_ptr->tm_gmtoff;
#else
	secs = cur_time - secs2;

	if(tm_ptr->tm_isdst) {
		/* adjust for daylight savings time */
		secs += 3600;
	}
#endif

	/* add in secs to make date based on Apple Jan 1, 1904 instead of */
	/*  Unix's Jan 1, 1970 */
	/* So add in 66 years and 17 leap year days (1904 is a leap year) */
	secs += ((66*365) + 17) * (24*3600);

	g_clk_cur_time = (word32)secs;

	clk_printf("Update g_clk_cur_time to %08x\n", g_clk_cur_time);
	g_clk_next_vbl_update = g_vbl_count + 5;
}

/* clock_update called by sim65816 every VBL */
void
clock_update()
{
	/* Nothing to do */
}

void
clock_update_if_needed()
{
	int	diff;

	diff = g_clk_next_vbl_update - g_vbl_count;
	if(diff < 0 || diff > 60) {
		/* Been a while, re-read the clock */
		update_cur_time();
	}
}

void
clock_write_c034(word32 val)
{
	g_c034_val = val & 0x7f;
	if((val & 0x80) != 0) {
		if((val & 0x20) == 0) {
			printf("c034 write not last = 1\n");
			/* set_halt(1); */
		}
		do_clock_data();
	}
}


void
do_clock_data()
{
	word32	mask;
	int	read;
	int	op;

	clk_printf("In do_clock_data, g_clk_mode: %02x\n", g_clk_mode);

	read = g_c034_val & 0x40;
	switch(g_clk_mode) {
	case CLK_IDLE:
		g_clk_read = (g_c033_data >> 7) & 1;
		g_clk_reg1 = (g_c033_data >> 2) & 3;
		op = (g_c033_data >> 4) & 7;
		if(!read) {
			/* write */
			switch(op) {
			case 0x0:	/* Read/write seconds register */
				g_clk_mode = CLK_TIME;
				clock_update_if_needed();
				break;
			case 0x3:	/* internal registers */
				g_clk_mode = CLK_INTERNAL;
				if(g_clk_reg1 & 0x2) {
					/* extend BRAM read */
					g_clk_mode = CLK_BRAM2;
					g_clk_reg1 = (g_c033_data & 7) << 5;
				}
				break;
			case 0x2:	/* read/write ram 0x10-0x13 */
				g_clk_mode = CLK_BRAM1;
				g_clk_reg1 += 0x10;
				break;
			case 0x4:	/* read/write ram 0x00-0x0f */
			case 0x5: case 0x6: case 0x7:
				g_clk_mode = CLK_BRAM1;
				g_clk_reg1 = (g_c033_data >> 2) & 0xf;
				break;
			default:
				halt_printf("Bad c033_data in CLK_IDLE: %02x\n",
					g_c033_data);
			}
		} else {
			printf("clk read from IDLE mode!\n");
			/* set_halt(1); */
			g_clk_mode = CLK_IDLE;
		}
		break;
	case CLK_BRAM2:
		if(!read) {
			/* get more bits of bram addr */
			if((g_c033_data & 0x83) == 0x00) {
				/* more address bits */
				g_clk_reg1 |= ((g_c033_data >> 2) & 0x1f);
				g_clk_mode = CLK_BRAM1;
			} else {
				halt_printf("CLK_BRAM2: c033_data: %02x!\n",
						g_c033_data);
				g_clk_mode = CLK_IDLE;
			}
		} else {
			halt_printf("CLK_BRAM2: clock read!\n");
			g_clk_mode = CLK_IDLE;
		}
		break;
	case CLK_BRAM1:
		/* access battery ram addr g_clk_reg1 */
		if(read) {
			if(g_clk_read) {
				/* Yup, read */
				g_c033_data = g_bram_ptr[g_clk_reg1];
				clk_printf("Reading BRAM loc %02x: %02x\n",
					g_clk_reg1, g_c033_data);
			} else {
				halt_printf("CLK_BRAM1: said wr, now read\n");
			}
		} else {
			if(g_clk_read) {
				halt_printf("CLK_BRAM1: said rd, now write\n");
			} else {
				/* Yup, write */
				clk_printf("Writing BRAM loc %02x with %02x\n",
					g_clk_reg1, g_c033_data);
				g_bram_ptr[g_clk_reg1] = g_c033_data;
				g_config_kegs_update_needed = 1;
			}
		}
		g_clk_mode = CLK_IDLE;
		break;
	case CLK_TIME:
		if(read) {
			if(g_clk_read == 0) {
				halt_printf("Reading time, but in set mode!\n");
			}
			g_c033_data = (g_clk_cur_time >> (g_clk_reg1 * 8)) &
									0xff;
			clk_printf("Returning time byte %d: %02x\n",
				g_clk_reg1, g_c033_data);
		} else {
			/* Write */
			if(g_clk_read) {
				halt_printf("Write time, but in read mode!\n");
			}
			clk_printf("Writing TIME loc %d with %02x\n",
				g_clk_reg1, g_c033_data);
			mask = 0xff << (8 * g_clk_reg1);

			g_clk_cur_time = (g_clk_cur_time & (~mask)) |
				((g_c033_data & 0xff) << (8 * g_clk_reg1));
		}
		g_clk_mode = CLK_IDLE;
		break;
	case CLK_INTERNAL:
		if(read) {
			printf("Attempting to read internal reg %02x!\n",
				g_clk_reg1);
		} else {
			switch(g_clk_reg1) {
			case 0x0:	/* test register */
				if(g_c033_data & 0xc0) {
					printf("Writing test reg: %02x!\n",
						g_c033_data);
					/* set_halt(1); */
				}
				break;
			case 0x1:	/* write protect reg */
				clk_printf("Writing clk wr_protect: %02x\n",
					g_c033_data);
				if(g_c033_data & 0x80) {
					printf("Stop, wr clk wr_prot: %02x\n",
						g_c033_data);
					/* set_halt(1); */
				}
				break;
			default:
				halt_printf("Writing int reg: %02x with %02x\n",
					g_clk_reg1, g_c033_data);
			}
		}
		g_clk_mode = CLK_IDLE;
		break;
	default:
		halt_printf("clk mode: %d unknown!\n", g_clk_mode);
		g_clk_mode = CLK_IDLE;
		break;
	}
}

